File: Java/QuizGame/src/QuizGame.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.stream.Stream;

import gameReport.Report;
import partite.MultiPlayer;
import partite.Partita;
import partite.SinglePlayer;
import utils.ConsoleCleaner;

public class QuizGame {
	//Due diversi path in base a se l'esecuzione parte da ide o da terminale
	public static String IDE_PATH = "..\\..\\Quiz\\";
	public static String TERMINAL_PATH = "..\\..\\..\\Quiz\\";

	public static void main(String[] args){
		ConsoleCleaner.clear();//Pulisce la console (solo su terminale)
		int scelta=0; //Comanda il flusso del software
		String quiz = null; //path del quiz selezionato
		
		Scanner in = new Scanner(System.in);
		
		Partita partita = null; //Partita da avviare
		Report risultato; //Report della partita 
		
		String quizDirPath;//Determinazione di quale path di base usare
		if(Paths.get("").toAbsolutePath().getFileName().toString().equals("src")){
			quizDirPath = TERMINAL_PATH;
		}else {
			quizDirPath = IDE_PATH;
		}
		
		do {
			//Inserimento modalità
			scelta = safeInput("Seleziona una modalita':"
					+ "\n1)Partita Singleplayer"
					+ "\n2)Partita Multiplayer"
					+ "\n0)Esci dal gioco"
					+ "\nInserisci una risposta valida (1-2-0): ",
					0, 2, in);
			
			//Calcolo della lista di quiz
			ArrayList<String> quizList = new ArrayList<>();
			
			try (Stream<Path> paths = Files.walk(Paths.get(quizDirPath))) {
			    paths
			        .filter(Files::isDirectory)
			        .map(Object::toString)
			        .forEach(quizList::add);
			} catch (IOException e) {
				System.out.println(e);
			}
			if(!quizList.isEmpty()) {
				quizList.remove(0);
			}
			in.nextLine();
			
			//Se scelta = 0 esco dall'esecuzione
			if(scelta == 0) {
				break;
			}
			
			//Inserimento del quiz da fare
			quiz = safeInput("Inserisci il nome del quiz che si vuole fare:"
					+ "\nQuiz disponibili: ", quizList, quizDirPath, in);
			
			//Esecuzione della modalità giusta
			if(scelta==1) {
				partita = new SinglePlayer(quiz);
			}
			else if(scelta == 2) {
				int players=safeInput("Inserisci il numero di giocatori (max 4): "
						, 1, 4, in);
				partita = new MultiPlayer(quiz, 
						players);
			}
			
			risultato = partita.start(in);
			
			//Stampa del report
			risultato.print();
			
			System.out.println("Premi invio per concludere la partita");
			in.nextLine();
			in.nextLine();
			ConsoleCleaner.clear();
			
			
		}while(scelta!=0);
		in.close();
	}

	
	//Definizione dei metodi per il safe input
	private static String safeInput(String out, ArrayList<String> choices, String prefix, Scanner in) {
		String res;
		do {
			System.out.println(out);
			
			choices.forEach(e ->{
				e=e.substring(prefix.length());
				System.out.println(e);
			});
			
			String choice = in.nextLine();
			res = prefix + choice;
			ConsoleCleaner.clear();
		}while(res==null || !choices.contains(res));
		return res;
	}

	private static int safeInput(String out, int min, int max, Scanner in) {
		int res;
		do {
			System.out.println(out);
			while(!in.hasNextInt()) {
				ConsoleCleaner.clear();
				System.out.println(out);
				in.next();
			}
			res = in.nextInt();
			ConsoleCleaner.clear();
		}while(res<min || res>max);
		
		return res;
	}
	
	
}

File: C++/QuizGame/src/QuizGame.cpp
#include <string>
#include <iostream>
#include <dirent.h>
#include <list>
#include <algorithm>
#include <memory>

#include "partite/SinglePlayer.h"
#include "partite/MultiPlayer.h"
#include "partite/SinglePlayerTimer.h"
using namespace std;

//Directory base dei quiz
const char* QuizDir = "../../Quiz/";

//Ritorna la lista di quiz disponibili
list<string> getQuizList();

//Permettono l'inserimento di valori in modo controllato
int safeInput(string out,int min, int max);
string safeInput(string out, list<string> choices);

int main() {

	system("CLS");
	int scelta=-1; //Comanda il flusso del programma
	int nGiocatori; //Per partita Multiplayer
	int seconds; //Per partita Timer
	string quiz;// Path del quiz

	unique_ptr<SinglePlayer> partita; //Oggetto partita

	while(true){
		scelta = safeInput("Seleziona una modalita':"
				"\n1)Partita Singleplayer"
				"\n2)Partita Multiplayer"
				"\n3)Partita Singleplayer a tempo"
				"\n4)Esci dal gioco\n",0,4);

		if(scelta==4)//Esci dall'applicazione
			break;

		//Inserimento del path del quiz che si vuole fare
		quiz = QuizDir+safeInput("Inserisci il quiz che si vuole fare: \n",getQuizList());

		//Avvio del tipo giusto di partita
		switch(scelta){
		case 1:
			partita= unique_ptr<SinglePlayer>(new SinglePlayer(quiz));
			partita->start();
			break;

		case 2:
			nGiocatori= safeInput("Inserisci il numero di giocatori (max 4): ",0,4);
			partita=unique_ptr<SinglePlayer>(new MultiPlayer(nGiocatori,quiz));
			partita->start();
			break;

		case 3:
			seconds= safeInput("Inserisci quanti secondi far durare la partita: ",1,3601);
			partita= unique_ptr<SinglePlayer>(new SinglePlayerTimer(seconds,quiz));
			partita->start();
			break;

		default:
			break;
		}
	}
	return 0;
}

//Implementazione metodi definiti all'inizio
int safeInput(string out, int min, int max){
	int res;
	cout<<out;
	cin>>res;
	while(!res || res<min || res> max){
		system("CLS");
		cout<<out;
		cin.clear();
		cin.ignore(40,'\n');
		cin>>res;
	}
	system("CLS");
	return res;
}

string safeInput(string out, list<string> choices){
	string res;
	do{
		cout<<out;
		for(auto choice : choices){
			cout<<choice<<"\n";
		}
		cin>>res;
		system("CLS");
		}while(find(choices.begin(), choices.end(), res) == choices.end()); //contains
	return res;
}

list<string> getQuizList(){
	DIR* dir_ptr;
	struct dirent *diread;
	list<string> content;

	if((dir_ptr = opendir(QuizDir)) != nullptr){
		while ((diread = readdir(dir_ptr)) != nullptr){
			content.push_back(diread->d_name);
		}
		closedir(dir_ptr);
	}
	content.remove(".");
	content.remove("..");
	return content;
}
